pep-0688.rst

```diff
@@ -0,0 +1,252 @@
+class implements the buffer protocol.
+
+For example, ``bytes`` may be declared as follows in a stub:
+
+.. code-block:: python
+
+   class bytes(types.Buffer, Sequence[int]):
+       def decode(self, ...): ...
+       ...
+
+The ``types.Buffer`` class does not require any special treatment
+in type checkers.
+
+Equivalent for older Python versions
+------------------------------------
+
+New typing features are usually backported to older Python versions
+in the ``typing_extensions`` package. Because the buffer protocol
+is accessible only in C, ``types.Buffer`` cannot be implemented
+in a pure Python package. As a temporary workaround, a
+``typing_extensions.Buffer`` ABC will be provided on Python versions
+that do not have ``types.Buffer`` available. For the benefit of
+static type checkers, ``typing_extensions.Buffer`` can be used as
+a base class in stubs to mark types as supporting the buffer protocol.
+For runtime uses, the ``ABC.register`` API can be used to register
+buffer classes with ``typing_extensions.Buffer``. When
+``types.Buffer`` is available, ``typing_extensions`` should simply
+re-export it.
+
+
+No special meaning for ``bytes``
+--------------------------------
+
+The special case stating that ``bytes`` may be used as a shorthand
+for other ``ByteString`` types will be removed from the ``typing``
+documentation.
+With ``types.Buffer`` available as an alternative, there is no good
+reason to allow ``bytes`` as a shorthand.
+We suggest that type checkers that implement this behavior should deprecate and
+eventually remove it.
+
+
+Backwards Compatibility
+=======================
+
+As the runtime changes in this PEP only add a new class, there are
+no backwards compatibility concerns.
+
+However, the recommendation to remove the special behavior for
+``bytes`` in type checkers does have backwards compatibility
+impact on users. An `experiment <https://github.com/python/mypy/pull/12661>`__
+with mypy shows that several major open source projects type
+checked with mypy will see new errors if the ``bytes`` promotion
+is removed. Nevertheless, the change improves overall type safety,
+so we believe the migration cost is worth it.
+
+
+Security Implications
+=====================
+
+None.
+
+
+How to Teach This
+=================
+
+We will add notes pointing to ``types.Buffer`` to appropriate places in the
+documentation, such as `typing.readthedocs.io <https://typing.readthedocs.io/en/latest/>`__
+and the `mypy cheat sheet <https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html>`__.
+Type checkers may provide additional pointers in their error messages. For example,
+when they encounter a place where a buffer object is passed to a function that
+is annotated to only accept ``bytes``, the error message could include a note suggesting
+to use ``types.Buffer`` instead.
+
+
+Reference Implementation
+========================
+
+[Link to any existing implementation and details about its state, e.g. proof-of-concept.]
```

Comment:
Perhaps remove the entire section until there's a reference implementation?
```suggestion
```

Reply:
But there is a reference implementation (after a lot of segfaults): https://github.com/python/cpython/compare/main...JelleZijlstra:typesbuffer?expand=1. I'll add a link to it.

Reply:
Ah, sorry! I'm only slowly getting half back into things, so I may have missed discussion on this in other fora.

A

Reply:
No, you were right! I only just pushed that implementation. Thanks for the review, I'm applying your suggestions right now.